website/preprints/tasks.py

```diff
@@ -18,29 +19,60 @@
 
 
 @celery_app.task(ignore_results=True, max_retries=5, default_retry_delay=60)
-def on_preprint_updated(preprint_id, update_share=True, share_type=None, old_subjects=None):
+def on_preprint_updated(preprint_id, update_share=True, share_type=None, old_subjects=None, saved_fields=None):
     # WARNING: Only perform Read-Only operations in an asynchronous task, until Repeatable Read/Serializable
     # transactions are implemented in View and Task application layers.
     from osf.models import PreprintService
     preprint = PreprintService.load(preprint_id)
     if old_subjects is None:
         old_subjects = []
-    if preprint.node:
+    if should_send_preprint_metadata(preprint, old_subjects, saved_fields):
         update_or_create_preprint_identifiers(preprint)
     if update_share:
         update_preprint_share(preprint, old_subjects, share_type)
 
+def should_send_preprint_metadata(preprint, old_subjects, saved_fields):
+    # Only call for a metadata update if identifiers haven't been requested yet, or subjects aren't being set
+    send_preprint_metadata = True
+    if not preprint.node:
+        send_preprint_metadata = False
+    if saved_fields and 'preprint_doi_created' in saved_fields:
+        send_preprint_metadata = False
+    if old_subjects:
+        send_preprint_metadata = False
+    return send_preprint_metadata
+
 def update_or_create_preprint_identifiers(preprint):
     status = 'public' if preprint.verified_publishable else 'unavailable'
     if preprint.is_published and not preprint.get_identifier('doi'):
-        get_and_set_preprint_identifiers(preprint)
+        request_identifiers(preprint)
     else:
         try:
             update_doi_metadata_on_change(preprint._id, status=status)
         except HTTPError as err:
             sentry.log_exception()
             sentry.log_message(err.args[0])
 
+def update_or_enqueue_on_preprint_updated(preprint_id, update_share=True, share_type=None, old_subjects=None, saved_fields=None):
+    task = get_task_from_postcommit_queue('website.preprints.tasks.on_preprint_updated', predicate=lambda task: task.kwargs['preprint_id'] == preprint_id)
+    if task:
+        old_subjects = old_subjects or []
+        task_subjects = task.kwargs['old_subjects'] or []
+        saved_fields = saved_fields or {}
+        task_saved_fields = task.kwargs['saved_fields'] or {}
+        saved_fields.update(task_saved_fields)
```

Comment:
Not sure if it matters, but this looks backwards. Shouldn't the old `saved_fields` be upated with the new `saved_fields`?